src/lib/Libwin/accesinfo.c

```diff
@@ -74,25 +74,29 @@ static ACL *
 create_secure_dacl(char *user, ACCESS_MASK mask, SID *owner_sid)
 {
 
-	DWORD					rids[1];
-	gid_t					grp[_MAX_GROUPS];
-	int						i, k;
-	int						cbAcl, cbAce;
-	ACL						*ndacl;
-	SID						*sid;
-	char					*name;
-	char	logb[LOG_BUF_SIZE] = {'\0' } ;
+	DWORD rids[1] = {0};
+	gid_t grp[_MAX_GROUPS] = {0};
+	int i = 0;
+	int k = 0;
+	int cbAcl = 0;
+	ACL *ndacl = NULL;
+	char logb[LOG_BUF_SIZE] = {'\0' };
 
 	rids[0] = DOMAIN_ALIAS_RID_ADMINS;
 	k = getgids(getlogin(), grp, rids);
 
 	if ((k < _MAX_GROUPS) && (owner_sid != NULL)) {
-		grp[k] = owner_sid;
+		grp[k] = sid_dup(owner_sid);
+		if (grp[k] == NULL) {
+			sprintf(logb, "failed to copy owner_sid");
+			log_err(-1, "create_secure_dacl", logb);
```

Comment:
We can use \_\_func\_\_ here.

Reply:
done

Reply:
can you explain what was happening here and what was wrong, and what is the updated one please?

Reply:
In end of this function we are doing
```
if (!EqualSid(grp[i], owner_sid))
        LocalFree(grp[i]);
```

Here we want to free all sid from grp array except one which is equal to owner_sid (passed as argument to this function), but we are using EqualSid which always returns equal is PBS service user (aka getlogin()) and user pointed owner_sid is same, means it will never free any sid from grp array.
To solve that, we create copy of owner_sid which is owned by this function and then unconditionally free every thing from grp at end.